### PR TITLE
Add element-rotation, in-place and lazy-wrapping, to collections

### PR DESCRIPTION
**Pull request by CTMacUser**
This change-set adds a new protocol requirement to `MutableCollection` to support in-place left-rotation and an extension for in-place left- and right-rotations.  It also adds a lazy wrapper for any `Collection` as a rotated view.

Resolves [SR-125](https://bugs.swift.org/browse/SR-125).

### CODE REVIEW

**Comment by CTMacUser** [comment]
I noticed on the GitHub pull request page that my text editors removed some trailing spaces in unrelated parts of `LazyCollection.swift` and `MutableCollection.swift`.  Sorry.  Can these be fixed?

**Comment by CTMacUser** [comment]
For the initial version, I copied `Reverse.swfit`, searched for that file name and "ReversedCollection" and did analogous changes.  I have no idea if that's the right way to add new functionality to the Standard Library.  I may be missing some other files to change.  Please advise.

File: stdlib/public/core/LazyCollection.swift [hunk 1]
```diff
@@ -10,23 +10,23 @@
...
-     return elements.lazy		
-   }		
- }		
-		
- extension LazyCollectionProtocol where Elements: LazyCollectionProtocol {		
-   // Lazy things are already lazy		
-   @inlinable // protocol-only		
-   public var lazy: Elements {		
-     return elements		
-   }
```
Comment (stephentyrone) [inline stdlib/public/core/LazyCollection.swift]: Try to keep unrelated whitespace changes out of patches.

File: stdlib/public/core/Rotate.swift [hunk 2]
```diff
@@ -0,0 +1,716 @@
...
+    ///   just before `endIndex`, unless it's `endIndex` (then it's ignored).
+    ///
+    /// - Returns: The new index of the value originally at `startIndex`.
+    ///
+    /// - Postcondition: The collection is a right-rotation of its old state,
+    ///   *i.e.* `newValue` equals `oldValue[middle>..] + oldValue[...middle]`.
+    ///
+    /// - Complexity: O(*n*), where *n* is the number of elements in the
+    ///   collection.
+    @inlinable
```
Comment (stephentyrone) [inline stdlib/public/core/Rotate.swift]: Why is this `@inlinable` and `rotate(toFirst:)` is not?

File: stdlib/public/core/Rotate.swift [hunk 3]
```diff
@@ -0,0 +1,716 @@
...
+    ///
+    /// - Complexity: O(*n*), where *n* is the number of elements in the
+    ///   collection.
+    @discardableResult
+    public mutating func rotate(toFirst middle: Index) -> Index {
+        switch middle {
+        case ...startIndex, endIndex...:
+            // A base case, otherwise the true/false and false/true cases of the
+            // switch below would trigger infinite recursion.
+            return startIndex
```
Comment (stephentyrone) [inline stdlib/public/core/Rotate.swift]: Why isn't it a precondition failure for `middle` to be an invalid index?

File: stdlib/public/core/Rotate.swift [hunk 4]
```diff
@@ -0,0 +1,716 @@
...
+            while true {
+                switch (i == middle, j == endIndex) {
+                case (false, false):
+                    // Move one step closer by doing the core swap.
+                    swapAt(i, j)
+                    formIndex(after: &i)
+                    formIndex(after: &j)
+                case (false, true):
+                    // The first partition is longer than the second.
+                    self[i...].rotate(toFirst: middle)
```
Comment (stephentyrone) [inline stdlib/public/core/Rotate.swift]: Swift doesn't have guaranteed TCO, and IIUC the bound on call depth here is potentially O(n), so this should really be written in terms of a loop rather than recursively.

Personally, I would write this something like the following sketch; I find it to be clearer than the state machine on (Bool,Bool), but different people may have different opinions on matters of taste.
```
while j != endIndex {
  while i < middle && j < endIndex {
    /* swap, update indices */
  }
  // update i, j, middle, result for next pass
}
return result
```

File: stdlib/public/core/Rotate.swift [hunk 5]
```diff
@@ -0,0 +1,716 @@
...
+    internal var _origin: Base.Index
+
+    /// Creates an instance that presents the elements of `base` in a rotated
+    /// order with the given index as the new start-index.
+    ///
+    /// - Complexity: O(1)
+    @inlinable
+    internal init(_base: Base, asFirst middle: Base.Index) {
+        self._base = _base
+        _origin = middle < _base.endIndex ? middle : _base.startIndex
```
Comment (stephentyrone) [inline stdlib/public/core/Rotate.swift]: Again, not convinced that out-of-bounds middle shouldn't be a precondition failure.

File: stdlib/public/core/Rotate.swift [hunk 6]
```diff
@@ -0,0 +1,716 @@
...
+        @inlinable
+        @inline(__always)
+        internal init(startingFrom middle: Base.Index) {
+            base = nil
+            _origin = middle
+        }
+        /// Creates an index into a rotated collection for the given base index,
+        /// with the other given index as the starting position.
+        ///
+        /// - Precondition: `base` cannot be the underlying collection's
```
Comment (stephentyrone) [inline stdlib/public/core/Rotate.swift]: This precondition should be enforced.

File: stdlib/public/core/Rotate.swift (hunk 2 above)
Comment (CTMacUser) [reply stdlib/public/core/Rotate.swift]: Because `rotate(toLast:)` is a trivial redirect to `rotate(toFirst:)`, but that doesn't mean that the latter's implementation is trivial.

File: stdlib/public/core/Rotate.swift (hunk 6 above)
Comment (CTMacUser) [reply stdlib/public/core/Rotate.swift]: That `Index` type can't know it.  The precondition has to be maintained from the client side (`RotatedCollection.rotation(of:)`, `RotatedCollection.endIndex`, and their clients within `RotatedCollection`).

File: stdlib/public/core/LazyCollection.swift (hunk 1 above)
Comment (CTMacUser) [reply stdlib/public/core/LazyCollection.swift]: Should be restored.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
